src/atlas_search.rs

```diff
@@ -0,0 +1,270 @@
+//! Helpers for building Atlas Search aggregation pipelines.
+
+mod gen;
+
+pub use gen::*;
+
+use std::marker::PhantomData;
+
+use crate::bson::{doc, Bson, Document};
+
+/// A helper to build the aggregation stage for Atlas Search.  Use one of the constructor functions
+/// and chain optional value setters, and then convert to a pipeline stage [`Document`] via
+/// [`into`](Into::into) or [`on_index`](AtlasSearch::on_index).
+///
+/// ```no_run
+/// # async fn wrapper() -> mongodb::error::Error {
+/// # use mongodb::{Collection, bson::{Document, doc}};
+/// # let collection: Collection<Document> = todo!()
+/// let cursor = coll.aggregate(vec![
+///     AtlasSearch::autocomplete("pre", "title")
+///         .fuzzy(doc! { "maxEdits": 1, "prefixLength": 1, "maxExpansions": 256 })
+///         .into(),
+///     doc! {
+///         "$limit": 10,
+///    },
+///    doc! {
+///        "$project": {
+///            "_id": 0,
+///            "title": 1,
+///         }
+///    },
+/// ]).await?;
+/// # Ok(())
+/// # }
+pub struct AtlasSearch<T> {
+    name: &'static str,
+    stage: Document,
+    _t: PhantomData<T>,
+}
+
+impl<T> Into<Document> for AtlasSearch<T> {
+    fn into(self) -> Document {
+        doc! {
+            "$search": {
+                self.name: self.stage
+            }
+        }
+    }
+}
+
+#[allow(missing_docs)]
+pub struct Built;
+
+impl<T> AtlasSearch<T> {
+    /// Finalize this builder.  Not typically needed, but can be useful to include builders of
+    /// different types in a single `Vec`.
+    pub fn build(self) -> AtlasSearch<Built> {
+        AtlasSearch {
+            name: self.name,
+            stage: self.stage,
+            _t: PhantomData::default(),
+        }
+    }
+
+    /// Like [`into`](Into::into), converts this builder into an aggregate pipeline stage
+    /// [`Document`], but also specify the search index to use.
+    pub fn on_index(self, index: impl AsRef<str>) -> Document {
+        doc! {
+            "$search": {
+                "index": index.as_ref(),
+                self.name: self.stage,
+            }
+        }
+    }
+}
+
+impl<T> IntoIterator for AtlasSearch<T> {
+    type Item = AtlasSearch<T>;
+
+    type IntoIter = std::iter::Once<AtlasSearch<T>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        std::iter::once(self)
+    }
+}
+
+/// Order in which to search for tokens.
+#[derive(Debug, Clone, PartialEq)]
+pub enum TokenOrder {
```

Comment:
Auxiliary types like this abstractly could fit better into the `gen` module since that's where they're used but since these don't actually depend on the input files it seemed better to hoist them up.